File: ios/Sources/ForgeIOSUI/ContainerRendererExtensions.swift
import SwiftUI
import ForgeIOSRuntime

public struct ForgeContainerRendererContext {
    public let runtime: ForgeRuntime?
    public let window: WindowContext?
    public let container: ContainerDef
    public let inheritedDataSourceRef: String?
    public let suppressTitle: Bool
    public let presentationDensity: ForgePresentationDensity
    public let targetContext: ForgeTargetContext?

    public init(
        runtime: ForgeRuntime?,
        window: WindowContext?,
        container: ContainerDef,
        inheritedDataSourceRef: String?,
        suppressTitle: Bool,
        presentationDensity: ForgePresentationDensity,
        targetContext: ForgeTargetContext?
    ) {
        self.runtime = runtime
        self.window = window
        self.container = container
        self.inheritedDataSourceRef = inheritedDataSourceRef
        self.suppressTitle = suppressTitle
        self.presentationDensity = presentationDensity
        self.targetContext = targetContext
    }
}

public protocol ForgeContainerRendererExtension {
    var kind: String { get }
    func render(context: ForgeContainerRendererContext) throws -> AnyView
}

public enum ForgeContainerRendererRegistryError: Error, Equatable {
    case emptyKind
    case duplicateKind(String)
}

public struct ForgeContainerRendererRegistry {
    private var renderers: [String: any ForgeContainerRendererExtension]

    public init(_ renderers: [any ForgeContainerRendererExtension] = []) throws {
        self.renderers = [:]
        for renderer in renderers {
            try register(renderer)
        }
    }

    public mutating func register(_ renderer: any ForgeContainerRendererExtension) throws {
        let kind = renderer.kind.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !kind.isEmpty else {
            throw ForgeContainerRendererRegistryError.emptyKind
        }
        guard renderers[kind] == nil else {
            throw ForgeContainerRendererRegistryError.duplicateKind(kind)
        }
        renderers[kind] = renderer
    }

    public func renderer(for kind: String?) -> (any ForgeContainerRendererExtension)? {
        guard let kind else { return nil }
        return renderers[kind.trimmingCharacters(in: .whitespacesAndNewlines)]
    }

    public var registeredKinds: Set<String> {
        Set(renderers.keys)
    }

    public static var empty: ForgeContainerRendererRegistry {
        try! ForgeContainerRendererRegistry()
    }
}

private struct ForgeContainerRendererRegistryKey: EnvironmentKey {
    static let defaultValue = ForgeContainerRendererRegistry.empty
}

extension EnvironmentValues {
    public var forgeContainerRendererRegistry: ForgeContainerRendererRegistry {
        get { self[ForgeContainerRendererRegistryKey.self] }
        set { self[ForgeContainerRendererRegistryKey.self] = newValue }
    }
}

File: android/sdk/src/main/java/com/viant/forgeandroid/ui/ContainerRendererExtensions.kt
package com.viant.forgeandroid.ui

import androidx.compose.runtime.Composable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import com.viant.forgeandroid.runtime.ContainerDef
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.ForgeTargetContext
import com.viant.forgeandroid.runtime.WindowContext

enum class ForgePresentationDensity {
    Automatic,
    Standard,
    Compact
}

data class ForgeContainerRendererContext(
    val runtime: ForgeRuntime,
    val window: WindowContext,
    val container: ContainerDef,
    val inheritedDataSourceRef: String?,
    val suppressTitle: Boolean,
    val presentationDensity: ForgePresentationDensity,
    val targetContext: ForgeTargetContext,
    val modifier: Modifier
)

fun interface ForgeContainerRendererExtension {
    @Composable
    fun Render(context: ForgeContainerRendererContext)
}

class ForgeContainerRendererRegistry private constructor(
    private val renderers: Map<String, ForgeContainerRendererExtension>
) {
    fun renderer(kind: String?): ForgeContainerRendererExtension? =
        kind?.trim()?.takeIf(String::isNotEmpty)?.let(renderers::get)

    val registeredKinds: Set<String>
        get() = renderers.keys

    class Builder {
        private val renderers = linkedMapOf<String, ForgeContainerRendererExtension>()

        fun register(
            kind: String,
            renderer: ForgeContainerRendererExtension
        ): Builder = apply {
            val normalized = kind.trim()
            require(normalized.isNotEmpty()) { "container renderer kind must not be blank" }
            require(!renderers.containsKey(normalized)) {
                "duplicate container renderer kind: $normalized"
            }
            renderers[normalized] = renderer
        }

        fun build(): ForgeContainerRendererRegistry =
            ForgeContainerRendererRegistry(renderers.toMap())
    }

    companion object {
        val Empty: ForgeContainerRendererRegistry = Builder().build()
    }
}

val LocalForgeContainerRendererRegistry = staticCompositionLocalOf {
    ForgeContainerRendererRegistry.Empty
}

val LocalForgePresentationDensity = staticCompositionLocalOf {
    ForgePresentationDensity.Automatic
}
